### PR TITLE
[DSA] Opt-in MXFP4 index-K cache for GLM-5.x / DeepSeek-V3.2 (port of #26209), with fused-kernel MXFP4 mode

### PR DESCRIPTION
**Pull request by vincentzed**
## Motivation

- The DSA (GLM-5.x / DeepSeek-V3.2) lightning-indexer K cache is stored as FP8: 128 B key + 4 B fp32 scale = **132 B/token/layer**.
- DeepSeek-V4 ships an opt-in MXFP4 indexer (#26209): E2M1 elements + packed UE8M0 group-32 scales = **68 B/token/layer**, with DeepGEMM FP4 MQA-logits kernels that measured 1.49–1.53× over FP8.
- This PR ports that to the shared DSA path behind `--enable-dsa-fp4-indexer` (default **off**): a ~48% cut of the indexer KV footprint, returned to the main KV pool as ~**+9.9% `max_total_num_tokens`**, plus the FP4 logits-kernel speedup.
- The fused indexer kernels (#27705, ~+10% single-stream) get a native **MXFP4 quant/store mode**, so on GLM-5.x (interleave) the flag keeps the fused path — FP4 is perf-neutral on top of fusion, instead of paying the ~8.5% split-path cost.
- Primary target/validation model: **nvidia/GLM-5.2-NVFP4**. DeepSeek-V3.2 (NeoX) is also supported: it takes the split indexer path under the flag (same as its FP8 default today); once the NeoX fusion fix (#30342) lands, the fused MXFP4 mode extends to it with a template knob.

## Modifications

1. **Server arg** `--enable-dsa-fp4-indexer` (default off) + fail-loud guards at `Indexer` construction: SM100+ only; DeepGEMM build must expose `fp8_fp4_paged_mqa_logits` / `fp8_fp4_mqa_logits` (there is no non-DeepGEMM FP4 fallback); DeepGEMM paged-MQA backend only; incompatible with `--enable-hisparse`, DSA prefill-CP, piecewise/breakable CUDA graph; `index_n_heads >= 32`.
2. **Pool** (`memory_pool.py`): `DSATokenToKVPool.index_k_bytes_per_token` = 128 (FP8) / 64 (MXFP4); rows sized `page_size * (index_k_bytes_per_token + 4)` (same block layout: K block, then per-token 4 B scales — fp32 for FP8, 4 packed UE8M0 exponents for FP4). `index_buf_accessor` gathers parametrized on the byte width; `pool_configurator` capacity estimate updated (that is what raises `max_total_num_tokens`).
3. **Routing** (`dsa_indexer.py`): under the flag, K stores go through the MXFP4 store (fused on interleave, split on NeoX), decode/verify top-k through `deep_gemm.fp8_fp4_paged_mqa_logits` over the 68 B paged view, prefill top-k through `deep_gemm.fp8_fp4_mqa_logits` over the gathered `(k_fp4, k_sf)` pair. Q is carried as the `(q_fp4, q_sf)` tuple the FP4 kernels consume; the per-token `q_scale` fold into the head-gate weights is dropped (the kernel applies `q_sf` itself).
4. **Fused kernels** (interleave path):
   - `csrc/deepseek_v32/indexer_k.cuh`: `kIsFp4` template mode on `fused_k_indexer_norm_rope_store` — LayerNorm + RoPE + per-32-elem-group amax (`warp::reduce_max<8>`) → UE8M0 ceil scale → packed E2M1 pairs into the 68 B page layout. Rounds through bf16 before quant and uses the Triton quantizer's scale order (`max(amax/6, 1e-4)`), so the output is **bit-identical** to the un-fused bf16-K-kernel + `store_fp4_index_k_cache` path.
   - `csrc/deepseek_v4/main_norm_rope.cuh`: `FusedQIndexerRopeHadamardFp4QuantKernel` generalized with the same `kRopeFirst/kHadamard` knobs as the FP8 Q kernel (defaults keep the V4 instantiation byte-compatible) plus a weight row stride for the non-contiguous `wk_weights_proj` slice; new rope-first (no-Hadamard) instantiation for DSA. `quant_fp4_e2m1` moves to the shared `sgl_kernel/deepseek_v4/fp8_utils.cuh`.
   - New JIT wrappers with distinct module cache keys: `fused_q_indexer_rope_first_fp4_quant`, `fused_k_indexer_norm_rope_store(..., is_fp4=)`.
5. **Tests** — `test/registered/jit/test_dsa_fp4_index_cache.py` (6 tests): quantizer format contract vs DeepGEMM's `per_token_cast_to_fp4(use_ue8m0=True, gran_k=32, use_packed_ue8m0=True)` (bit-equal); store→gather round-trip on the 68 B layout (bit-exact, scattered locs); FP4-vs-FP8 logits top-2048 index-set overlap (ragged + paged + ragged↔paged consistency, SM100-gated); fused K FP4 store byte-exact vs un-fused (strided inputs); fused Q FP4 vs eager reference (>99.9% byte match, exact head-gate weights).

Not changed:

- Flag off: the FP8 path is byte-identical (132 B sizing and kernel dispatch untouched).
- The fusion gate (`is_neox_style` clause) and the split path.
- The DeepSeek-V4 native indexer (shared kernels extended with defaulted template args; `test/registered/jit/deepseek_v4/test_fp4_indexer.py` 12/12).
- No NVFP4/E4M3 mode — the DeepGEMM sm100 MQA-logits kernel hardcodes E2M1 + UE8M0; this is MXFP4 only.
- Index-share / `skip_topk` logic; spec decode is wired through the same FP4 paged call (next_n≥2 shapes) but not yet validated E2E.

Cache reduction at fixed `--mem-fraction-static 0.85` (startup `max_total_num_tokens`, B300 ×4, tp4 ep4):

| model | FP8 index | MXFP4 index | delta |
|---|---|---|---|
| nvidia/GLM-5.2-NVFP4 | 2,292,288 | 2,519,680 | **+9.92%** |
| deepseek-ai/DeepSeek-V3.2 | 1,651,392 | 1,815,488 | **+9.94%** |

Exactly the layout prediction `(576+132)/(576+68) = 1.0994`.

## Accuracy Tests

GSM8K full set, `python3 -m sglang.test.few_shot_gsm8k --num-questions 1319 --num-shots 20`, `/flush_cache` between runs, B300 ×4, `--tp 4 --ep 4 --disable-radix-cache`, fused indexer path in both arms:

```
nvidia/GLM-5.2-NVFP4, fused FP8 index (baseline):
Accuracy: 0.945   Accuracy: 0.942   Accuracy: 0.945   Accuracy: 0.937   Accuracy: 0.944   mean 0.943

nvidia/GLM-5.2-NVFP4, fused MXFP4 index (this PR, flag on):
Accuracy: 0.936   Accuracy: 0.939   Accuracy: 0.947   Accuracy: 0.948   Accuracy: 0.939   mean 0.942
```

- Flat (−0.08 pt, distributions fully overlapping) — better than the ~1.5 pt envelope #26209 measured on V4-Flash.
- The fused FP4 path drops the Hadamard rotation exactly like fused FP8 does; the flat result shows that costs nothing measurable.

AIME25 (harder signal) — `sgl-eval run aime25 --n-repeats 8 --max-tokens 65536 --temperature 1.0 --top-p 0.95 --num-threads 128`, same launch, sequential arms:

nvidia/GLM-5.2-NVFP4 (fused path both arms):

```
FP8 index   : pass@1[avg-of-8] = 92.08% +/- 3.05% (SEM 1.08%)   pass@8 100.00%   majority@8 95.00%   no_answer 1.25%
MXFP4 index : pass@1[avg-of-8] = 91.25% +/- 3.96% (SEM 1.40%)   pass@8 100.00%   majority@8 96.67%   no_answer 1.25%
```

deepseek-ai/DeepSeek-V3.2 (split path under the flag):

```
FP8 index   : pass@1[avg-of-8] = 88.75% +/- 1.73% (SEM 0.61%)   pass@8 93.33%   majority@8 90.00%
MXFP4 index : pass@1[avg-of-8] = 91.25% +/- 2.48% (SEM 0.88%)   pass@8 93.33%   majority@8 93.33%
```

- Parity within noise on both models (GLM -0.83 pt, V3.2 nominally +2.5 pt; SEM ~1 pt); wall-clock equal across arms.

DeepSeek-V3.2 (NeoX → split path under the flag), GSM8K ×5, radix on:

```
FP8 index : Accuracy: 0.955   Accuracy: 0.958   Accuracy: 0.957   Accuracy: 0.955   Accuracy: 0.959   mean 0.957
MXFP4 index: Accuracy: 0.956   Accuracy: 0.957   Accuracy: 0.958   Accuracy: 0.960   Accuracy: 0.957   mean 0.958
```

Unit tests:

```
python3 -m pytest test/registered/jit/test_dsa_fp4_index_cache.py -q
6 passed, 21 warnings in 10.58s

python3 -m pytest test/registered/jit/test_dsv32_indexer_fusion.py -q
7 passed, 21 warnings in 7.13s

python3 -m pytest test/registered/jit/deepseek_v4/test_fp4_indexer.py -q   # shared main_norm_rope.cuh, V4 path
12 passed, 7 warnings in 6.67s
```

## Speed Tests and Profiling

Kernel-level, DeepGEMM paged MQA-logits FP8 vs FP4 (B300, B=256, H=64, D=128, NextN=1, block_kv=64; 50 iters, CUDA-event timed):

| case | FP8 µs | FP4 µs | FP4 speedup |
|---|---|---|---|
| B=256, L=8192 | 49.01 | 29.99 | **1.63×** |
| B=256, L=32768 | 170.54 | 97.13 | **1.76×** |

(#26209 measured 1.49× / 1.53× for the same cases on B200.)

GLM-5.2-NVFP4 GSM8K wall-clock, same servers as the accuracy runs (5 runs/arm, fused path both arms): FP8 **186.9 s / 759 tok/s** vs MXFP4 **187.0 s / 758 tok/s** — **neutral on top of fusion**. (An earlier revision that forced the split path under the flag cost ~8.5% here; the fused MXFP4 kernels recover all of it.)

AIME25 wall-clock was equal across arms on both models.

> During validation we briefly hit a long-decode instability; it turned out to be the fused top-k v2 tie-overflow already fixed upstream by #30512, and all numbers in this PR were measured with that fix included.

Refs: #26209 #27705 #30111 #30342 #30378

---
### CI States

Latest PR Test (Base): :x: [Run #28958182240](https://github.com/sgl-project/sglang/actions/runs/28958182240)
Latest PR Test (Extra): :x: [Run #28958182103](https://github.com/sgl-project/sglang/actions/runs/28958182103)

### CODE REVIEW

**Comment by vincentzed** [comment]
@b8zhong
